File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.py
class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        def solve(n, target):
            nonlocal k
            if target==0 and n==0: return 1
            if target<=0 or n<=0 : return 0
            if dp[target][n]!=-1: return dp[target][n]
            
            ans, lim = 0, min(k, target)+1
            for i in range(1, lim):
                ans+=solve(n-1, target-i)
            dp[target][n]=ans
            return dp[target][n]
        
        dp=[[-1]*(n+1) for i in range(target+1)]
        return solve(n, target)%(10**9+7)

File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.java
class Solution {
    int MOD = 1_000_000_007;
    Integer[][] memo = new Integer[31][1001];
    public int numRollsToTarget(int n, int k, int target) {
        if(n==0) return target==0 ? 1 : 0;
        
        if(memo[n][target] != null) return memo[n][target];
        
        memo[n][target] = 0;
        
        for(int x=1; x<=k && x<=target; ++x) 
            memo[n][target] = (memo[n][target] + numRollsToTarget(n-1, k, target-x))%MOD;
        
        return memo[n][target];
    }
}


// // Recursion
// // O(k^n)
//
// class Solution {
//     int MOD = 1_000_000_007;
//     public int numRollsToTarget(int n, int k, int target) {
//         if(n==0) return target==0 ? 1 : 0;
//         int ans = 0;
        
//         for(int x=1; x<=k && x<=target; ++x) ans = (ans + numRollsToTarget(n-1, k, target-x))%MOD;
//         return ans;
//     }
// }